### PR TITLE
Lesson 4. Enhance intro; add mini-quizzes; final quiz

### PR DESCRIPTION
**Pull request by elPiablo**
Lesson 4 teaches us how to write automated tests for our smart contracts
PR aims to give a little enhancement to intro; add final quiz and add mini-quizzes to yet another amazing D_D home grown lesson from @briangershon 

Just going to use the following, new template to check everything off (good practices for new Members to see), despite most stuff already having been created.

### Empowering Web3 Developers: **Guidelines for Creating High-Quality Tutorials and Content**

Lesson Components:

- [x] A meaningful title
- [x] Intro paragraph to clarify assumptions, and build confidence
  - Introduce learners' capabilities, competencies and experience, including proficiency with developer tools.
  - Clearly state intended learning outcome, referring to relevant previous lessons and setting context for subsequent lessons.
  - Offer estimated minimum and maximum completion times for lesson, catering to learners' respective competencies.
  - Learning extends far beyond the 'what' and 'how'. Provide the essential 'why', by showcasing real-world examples of how new knowledge can be used for social-environmental impact.
- [x] 2nd paragraph to introduce our pedagogy and lesson structure
  - We assess prior knowledge and test learning through mini-quiz checkpoints.
  - We engage learners with prediction activities and a final quiz. Guide them to join the forum.
  - We embrace well-being by encouraging regular breaks.
- [x] Warmup Quiz 1
- [x] List of contents and relevant tools.
- [x] A lead-in to lesson.
- [x] Lesson chunk 
  - Meaningful digestible chunks align with learners' attention span.
  - Each chunk focuses on specific learning concepts for streamlined understanding.
  - A mini-quiz checkpoint of ~5 multiple choice questions reinforces retention.
  - A chance for a regular break, conducive to optimal learning and a culture of well-being.
- [x] Checkpoint quiz 2
- [x] Lesson chunk
- [x] Checkpoint quiz 3
- [x] Lesson chunk 
- [x] Checkpoint quiz 4
- [x] Lesson chunk 
- [x] Checkpoint quiz 5
- [x] Lesson chunk 
- [x] Checkpoint quiz 6 (quite sure we won't need this one)
(Additional lesson chunks and their corresponding checkpoint quizzes as needed)

- [x] A final quiz of ~12 multiple choice questions for overall comprehension.
- [x] Brief description of upcoming lesson to consolidate the learner’s path.
- [x] Kudos to learner, and prompt to engage in forum with peers.

### CODE REVIEW

**Comment by elPiablo** [comment]
Don't worry folks, the questions/answers in there at this stage are only placeholders, so the code can run without issues

**Comment by elPiablo** [comment]
Hey @briangershon and @wolovim I've done three commits for this PR so far, but it's not actually a tonne of work.... 

First was to simply create a file/directory structure where I 'pre-filled' all the .json files and their placeholders in the .mdx, so the code wouldn't break i.e. the questions and answers were irrelevant to lesson 4. 

Second was to add a mini-quiz for the second part i.e. I skipped the warm-up quiz for now and went to the next checkpoint questions

Third was to come back and delete all the pre-filled text I had initially added to the .json files, because I didn't want to cause any confusion when you're reviewing, by seeing Qs and As that aren't relevant. I left all the syntax in there as multiple choice  templates, but erased all the keys and values. So it should be a lot easier to read for you. (should have done this in the first commit).

Needless to say, if you're in the mood, fill some Qs and As in while you're there!

File: utils/quizzes/lesson-4-quiz.json
Comment (briangershon) [inline utils/quizzes/lesson-4-quiz.json:20]: I wonder if one of the final questions should be pulled from the TokenURI and SVG sections -- namely discussions of strategies for breaking up the logic to make it more testable?

And maybe a question about TDD in the outro?

(these would replace this general Ethereum questions and the following "What can we use a smart contract for?" questions)

**Review by briangershon — COMMENTED** [review]
Thanks @elPiablo for working on adding the quizzes!

I added a couple of comments, though didn't have a chance to try to start filling out the quiz sections that don't have questions yet.

File: utils/quizzes/lesson-4-quiz.json
Comment (elPiablo) [reply utils/quizzes/lesson-4-quiz.json:20]: My bad @briangershon. I should have emptied the .json files of all this text too (like I did with the mini-quiz questions last night). This the pre-fill text/questions from https://quiz-editor.vercel.app/ (you can always find this in /utils/quiz-creation-tools.txt)

File: utils/questions/lesson-4/2-get-set-up/Q1.json
Comment (elPiablo) [reply utils/questions/lesson-4/2-get-set-up/Q1.json:14]: Of course @briangershon. I had thought of putting "Only a) and c)" as the last line and "All of the above" as second last line, but it looked funny - but would have actually made sense. I was trying to trip up the reader with this one, by testing if they had actually digested the "Help write modular code" information.

**Comment by elPiablo** [comment]
Hardhat and ethers.js have had an upgrade, all in here:
https://github.com/NomicFoundation/hardhat/releases/tag/%40nomicfoundation%2Fhardhat-toolbox%403.0.0
So the lesson needed some upgrades.
I only managed to get these changes in. 

Lesson at this stage is still not in an M **viable** P state. Need to comment some things out, e.g. placeholders for quizzes,  before it is **viable** for merging.

**Review by elPiablo — COMMENTED** [review]
Hey @briangershon and @wolovim 

I did a few commits (manageable chunks) in the last 24 hours. I forgot about explicitly notifying you. I'm doing that now after selecting the PR, selecting Files changed and now clicking on the green Review changes drop down, and selecting Comment.

I finished the mini-quizzes from my side. Now we need your expertise to check them, change them, or replace them with more appropriate versions - you lads are the gurus here. All hail!!

If there is some improvement I can make in my workflow, please let me know. 
Cheers!

**Review by elPiablo — COMMENTED** [review]
@briangershon and @wolovim I added the final quiz. 

I was sort of building up the final quiz (more comprehension of concepts questions) as I was going along while creating the mini-quizzes (more specific/targeted information questions). Different types of memory retention needed from learner.

But I didn't get round to check against the mini-quizzes yet for blatant repetition. 

Feel free to change anything that simply isn't right.

**Comment by okhaimie-dev** [comment]
@elPiablo 

I am curious where I can find the warm up quiz data for lesson 4. 

```tsx 
<SideDrawer buttonText="Checkpoint Questions" title="Refactoring Code and Testing SVG">
      <Question question="lesson-4/5-SVG-test/Q1" />
      <Question question="lesson-4/5-SVG-test/Q2" />
      <Question question="lesson-4/5-SVG-test/Q3" />
      <Question question="lesson-4/5-SVG-test/Q4" />
      <Question question="lesson-4/5-SVG-test/Q5" />
</SideDrawer>
```
Location: `pages/lessons/projects/4.mdx`

TBH, The first and last warm up question is unclear, and I think we can improve it.

CleanShot 2023-07-18 at 09 26 43@2x

**Comment by elPiablo** [comment]
@okhaimie-dev the actual jsons with the questions and final-quizzes are in the `utils` directory e.g. `/utils/questions/lesson-4/1-intro/Q5-json for that one you're referring to (on branch `quiz-lesson-4`)
There's another files in there called `/utils/quiz-creation-tools.txt` for an explainer and  you can have a look at the cool quiz creation app @0xChin built!!

**Review by elPiablo — COMMENTED** [review]
@briangershon @wolovim I made a start on the intro in Lesson 4. 

All your original wording is in there, Brian. Just changed the bullet points into a paragraph and added a few new bits, of which I tried a bit to get your tone. If you're up for re-writing it instead, cool!

That's it for now.

File: pages/lessons/projects/4.mdx
Comment (wolovim) [inline pages/lessons/projects/4.mdx:31]: are these `PART #`'s for the sake of your review process or are they intended to remain in the text? based on past PRs, i'm assuming the former and that they need removing prior to shipping.

**Review by elPiablo — COMMENTED** [review]
@briangershon and @wolovim I finished my update of the lesson. Hope I did it justice, Brian 😊

**Review by wolovim — APPROVED** [review]
Thanks for another major investment 🙏 

Same process as the last couple lessons: just gave it a quick sanity check, made a few quality nitpicks that stood out, and am approving now so that future improvements can happen in follow-up PRs.

**Pull request merged**
